File: src/HmacManager/Components/HmacProvider.cs
using HmacManager.Extensions;
using HmacManager.Headers;

namespace HmacManager.Components;

public class HmacProvider : IHmacProvider
{
    private readonly HmacProviderOptions _options;
    private readonly ContentHashGenerator _contentHashGenerator;
    private readonly SignatureHashGenerator _signatureHashGenerator;

    public HmacProvider(
        HmacProviderOptions options,
        ContentHashGenerator contentHashGenerator,
        SignatureHashGenerator signatureHashGenerator
    )
    {
        _options = options;
        _contentHashGenerator = contentHashGenerator;
        _signatureHashGenerator = signatureHashGenerator;
    }

    public Task<string> ComputeSignatureAsync(string signingContent)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(signingContent);

        return _signatureHashGenerator.HashAsync(signingContent);
    }

    public async Task<string> ComputeSigningContentAsync(
        HttpRequestMessage request, 
        DateTimeOffset dateRequested, 
        Guid nonce,
        HeaderValue[]? headerValues = null
    )
    {
        var builder = new SigningContentBuilderValidated(request)
            .WithPublicKey(_options.Keys.PublicKey)
            .WithDateRequested(dateRequested)
            .WithNonce(nonce)
            .WithHeaderValues(headerValues ?? []);

        if (request.TryGetContent(out var content))
        {
            var contentString = await content.ReadAsStringAsync();
            var contentHash = await _contentHashGenerator.HashAsync(contentString);
            builder.WithContentHash(contentHash);
        }

        return builder.Build();
    }
}

File: src/Scripts/HmacManager/components/hmac-provider.ts
import { SignatureBuilder } from "../builders/signature-builder.js"
import { SigningContentBuilder } from "../builders/signing-content-builder.js"

export class HmacProvider {
    readonly publicKey: string;
    readonly privateKey: string;
    readonly signedHeaders: string[];

    constructor(publicKey, privateKey, signedHeaders) {
        this.publicKey = publicKey
        this.privateKey = privateKey
        this.signedHeaders = signedHeaders
    }

    computeSigningContent = (request: Request, dateRequested: Date, nonce: string): string => {

        const signingContent = new SigningContentBuilder(request)
            .withPublicKey(this.publicKey)
            .withDateRequested(dateRequested)
            .withNonce(nonce)
            .withHeaderValues(this.signedHeaders)
            //.withContentHash() // TODO
            .build()

        return signingContent
    }

    computeSignature = async (signingContent: string): Promise<string> => {

        const signaturePromise = new SignatureBuilder()
            .withPrivateKey(this.privateKey)
            .withSigningContent(signingContent)
            .build()

        return signaturePromise
    }
}